sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts

```diff
@@ -232,6 +242,99 @@ export class AssessmentController {
+        jobLogger.error(message);
+        return job.error(ASSESSMENT_NOT_FOUND, message);
+      }
+      const assessmentId = job.variables.assessmentId;
+      const studentId = assessment.application.student.id;
+      const programYearId = assessment.application.programYear.id;
+      jobLogger.log(
+        `Verifying the assessment calculation order for assessment: ${assessmentId} student:${studentId} program year: ${programYearId}.`,
+      );
+      // Check for any assessment with ongoing calculation.
+      const assessmentInCalculationStep =
+        await this.studentAssessmentService.getAssessmentInCalculationStepForStudent(
+          studentId,
+          programYearId,
+        );
+
+      // If an ongoing calculation is happening all other assessments for given student in program year
+      // must wait until the calculation is complete and saved to the system.
+      // Also if a worker job is terminated by any chance after updating the calculation start date
+      // for an assessment, then the assessment must proceed for the calculation.
+      if (
+        assessmentInCalculationStep &&
+        assessmentInCalculationStep.id !== assessmentId
+      ) {
+        jobLogger.log(
+          `There is ongoing calculation happening for assessment: ${assessmentInCalculationStep.id} ` +
+            `and hence assessment : ${assessmentId} is waiting for that calculation to complete.`,
+        );
+        return job.complete({ isReadyForCalculation: false });
+      }
+      jobLogger.log(
+        `There is no ongoing calculation happening while verifying the order for assessment: ${assessment}.`,
+      );
+      // Get the first outstanding assessment waiting for calculation as per the sequence.
+      const [firstOutstandingStudentAssessment] =
+        await this.studentAssessmentService.getOutstandingAssessmentsForStudentInSequence(
+          studentId,
+          programYearId,
+        );
+      jobLogger.log(
+        `The first outstanding assessment is identified to be assessment: ${firstOutstandingStudentAssessment.id} ` +
+          `with original assessment start date: ${firstOutstandingStudentAssessment.originalAssessmentStudyStartDate} ` +
+          `while verifying the order for assessment: ${assessmentId}.`,
+      );
+      // If the processing assessment is same as first outstanding assessment
+      // then proceed for calculation.
+      const isReadyForCalculation =
+        firstOutstandingStudentAssessment.id === job.variables.assessmentId;
+      if (isReadyForCalculation) {
+        await this.studentAssessmentService.saveAssessmentCalculationStartDate(
+          assessmentId,
+        );
+      }
+      jobLogger.log(
+        `The assessment calculation order has been verified and ready for calculation status is ${isReadyForCalculation} ` +
+          `for assessment: ${assessmentId}.`,
+      );
+      return job.complete({ isReadyForCalculation });
```

Comment:
I'm just wondering if it is possible that `isReadyForCalculation` could be false and we aren't handling the case for `isReadyForCalculation = false` here.

Reply:
We are good here for the condition `isReadyForCalculation = false` here @lewischen-aot . The action happening in line 322 is only meant for `isReadyForCalculation = true` and when it is false then variable must be returned as it is with job complete.

Reply:
Thanks @dheepak-aot This makes now.